### PR TITLE
feat: add the ability to trace gas consumption

### PR DESCRIPTION
**Pull request by evan-forbes**
## Overview

bringing back the gas monitor, not sure if we should merge this, but I'm using to regenerate all the gas charts, which can be found in #2132 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: tools/gasmonitor/gas_monitor.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+// ConsumeGas wraps the underlying RefundGas method to record the amount of gas
+// refunded.
+func (gm *GasConsumptionTrace) RefundGas(amount sdk.Gas, descriptor string) {
+	gm.Readings = append(gm.Readings, Reading{Description: descriptor, Refund: true, Amount: amount})
+	gm.GasMeter.RefundGas(amount, descriptor)
+}
+
+// AnteHandle replaces the gas meter with the provided context with one that
+// monitors the amount of gas consumed it fulfills the ante.Decorator interface.
```
Comment (rootulp) [inline tools/gasmonitor/gas_monitor.go:110]:
```suggestion
// AnteHandle replaces the gas meter in the provided context with one that
// monitors the amount of gas consumed. It fulfills the ante.Decorator interface.
```

File: tools/gasmonitor/integration_test.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			signer := blobtypes.NewKeyringSigner(s.cctx.Keyring, s.accounts[count], s.cctx.ChainID)
+			res, err := blob.SubmitPayForBlob(context.TODO(), signer, s.cctx.GRPCClient, tc.blobs, tc.opts...)
+			require.NoError(t, err)
+			require.NotNil(t, res)
+			require.Equal(t, abci.CodeTypeOK, res.Code, res.Codespace, res.Logs)
+		})
+		count++
+	}
```
Comment (rootulp) [inline tools/gasmonitor/integration_test.go:257]: [optional refactor] Is it possible to remove the `count` variable entirely?

```suggestion
	for i, tc := range tests {
		s.Run(tc.name, func() {
			signer := blobtypes.NewKeyringSigner(s.cctx.Keyring, s.accounts[i], s.cctx.ChainID)
			res, err := blob.SubmitPayForBlob(context.TODO(), signer, s.cctx.GRPCClient, tc.blobs, tc.opts...)
			require.NoError(t, err)
			require.NotNil(t, res)
			require.Equal(t, abci.CodeTypeOK, res.Code, res.Codespace, res.Logs)
		})
	}
```

**Review by rootulp — DISMISSED** [review]
Overall LGTM. Nothing in the PR seems particularly difficult to maintain and the Python script is a separate Gist so it seems safe to merge but I'm curious what other reviewers think.

File: tools/gasmonitor/integration_test.go (hunk 2 above)
Comment (evan-forbes) [reply tools/gasmonitor/integration_test.go:257]: yeah good catch forgot about that lil guy

File: tools/gasmonitor/integration_test.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+			},
+		},
+		{
+			"small random typical",
+			blobfactory.ManyRandBlobs(t, tmrand.NewRand(), 1, 1),
+			[]blobtypes.TxBuilderOption{
+				blobtypes.SetFeeAmount(sdk.NewCoins(sdk.NewCoin(app.BondDenom, sdk.NewInt(9000)))),
+				blobtypes.SetGasLimit(90_000),
+			},
+		},
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply tools/gasmonitor/integration_test.go:228]: yeah exactly, I think I just forgot to fix the names, will go do now

**Review by rootulp — APPROVED** [review]

File: tools/gasmonitor/integration_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply tools/gasmonitor/integration_test.go]: https://github.com/celestiaorg/celestia-app/pull/2131/commits/6ac9df335bbcaa63048562b77f70103215f82456

**Comment by cmwaters** [comment]
> Do we see ourselves tracing the consumption in the future? If not, we should close again. Opening the discussion to the reviewers.

I guess do we see the calculations of gas being altered in the future (I'm not sure if an application can even set the `GasConfig` currently).

**Comment by evan-forbes** [comment]
something like this might be useful if we ever run into another gas bug, or if we ever want to fine tune resource pricing. but until then... this is a lot of extra code!

**Pull request closed without merging**